### PR TITLE
OpenJSSE Platform support

### PR DESCRIPTION
**Pull request by yschimke**
Support for HTTP/2 with https://github.com/openjsse/openjsse

Caveats

1) Tests based on MockWebserver (on JDK8) won't work without larger changes
2) Unclear what the overall status is of the project is, and edge cases like cipher selections on older VMs may need testing.

### CODE REVIEW

**Comment by yschimke** [comment]
This seems cool, but could be too early for us.

**Comment by yschimke** [comment]
n.b. Failures relate to weird mix of jdk8 and TLSv1.3.  Not a bug in this PR per se, just something to sidestep more carefully than we are currently.

```
Caused by: java.lang.IllegalArgumentException: TLSv1.3
	at sun.security.ssl.ProtocolVersion.valueOf(ProtocolVersion.java:187)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:206)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
	at org.openjsse.sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1308)
	at org.openjsse.sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1199)
	at org.openjsse.sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1146)
```

**Comment by yschimke** [comment]
image

**Comment by yschimke** [comment]
@swankjesse Getting MockWebserver working with openjsse is a bigger effort.  Options

1) We could test twice.  Basic client only tests on JDK8, test client+server with JDK11? 
2) Fix MockWebserver completely to handle openjsse on JDK8.

**Comment by swankjesse** [comment]
What’s difficult about getting MockWebServer working with OpenJSSE?

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * Platform using OpenJSSE (https://github.com/openjsse/openjsse) if installed as the first
+ * Security Provider.
+ *
+ * Requires org.openjsse:openjsse >= 1.1.0 on the classpath.
+ */
+class OpenJSSEPlatform private constructor() : Platform() {
+  private val provider: Provider = org.openjsse.net.ssl.OpenJSSE()
+
+  override fun newSSLContext(): SSLContext =
+      SSLContext.getInstance("TLSv1.3", provider)
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: is this TLSv1.3+ ? We also wanna support TLSv1.2 right?

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    val trustManagers = factory.trustManagers!!
+    check(trustManagers.size == 1 && trustManagers[0] is X509TrustManager) {
+      "Unexpected default trust managers: ${trustManagers.contentToString()}"
+    }
+    return trustManagers[0] as X509TrustManager
+  }
+
+  public override fun trustManager(sslSocketFactory: SSLSocketFactory): X509TrustManager? =
+      throw UnsupportedOperationException(
+          "clientBuilder.sslSocketFactory(SSLSocketFactory) not supported with OpenJSSE")
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: nice

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    if (sslSocket is org.openjsse.javax.net.ssl.SSLSocket) {
+      val sslParameters = sslSocket.sslParameters
+
+      if (sslParameters is org.openjsse.javax.net.ssl.SSLParameters) {
+        // Enable ALPN.
+        val names = alpnProtocolNames(protocols)
+        sslParameters.applicationProtocols = names.toTypedArray()
+      }
+
+      sslSocket.sslParameters = sslParameters
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: what does this do?

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    } else {
+      super.configureTlsExtensions(sslSocket, hostname, protocols)
+    }
+  }
+
+  override fun getSelectedProtocol(sslSocket: SSLSocket): String? =
+      if (sslSocket is org.openjsse.javax.net.ssl.SSLSocket) {
+        when (val protocol = sslSocket.applicationProtocol) {
+          null, "" -> null
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: does it actually return both of these values? ick

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt (hunk 3 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: javax.net.ssl.SSLSocket creates a new params object when you use the getter, so to apply you specifically need to call the setter.

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt (hunk 4 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: I don't make the rules...

```
     * @return null if it has not yet been determined if application
     *         protocols might be used for this connection, an empty
     *         {@code String} if application protocols values will not
     *         be used, or a non-empty application protocol {@code String}
     *         if a value was successfully negotiated.
```

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt (hunk 3 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: For OpenJSSE it at least triggers changes to packet sizes https://github.com/openjsse/openjsse/blob/4b67b70d1ffe28f62f388e9a4cd7768be23b5f6c/src/main/java/org/openjsse/sun/security/ssl/SSLSocketImpl.java#L1048

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt (hunk 1 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: It does https://github.com/openjsse/openjsse/blob/82abe9803bbae50428be27189ab85b48d20f96bd/src/main/java/org/openjsse/sun/security/ssl/SSLContextImpl.java#L808

But in JDKs, supported vs defaulted often diverge so on other OpenJDK/Oracle/Android platforms we have had to select TLSv1.2 specifically or we might get held back.

**Comment by yschimke** [comment]
> What’s difficult about getting MockWebServer working with OpenJSSE?

A mixture of JVM singletons (TlsUtils.localhost), provider setup timing, strict code in the JDK that throws exceptions if it sees the 1.3 version being selected, assumptions in MockWebServer e.g. it doesn't use the Platform SSLContext API.  So it's a separate larger task to fix it up.  Work we should do, but doesn't block using this, and we can already start testing this as a client only where it appears to mostly *just* work.

**Comment by yschimke** [comment]
Raised https://github.com/square/okhttp/issues/5380 to fix up MockWebserver.  But not blocking this as we can test thoroughly with client only to flush out bugs before we bake the 4.2 cake.

File: okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/platform/OpenJSSEPlatform.kt]: Got it. APIs that mix mutable and accessors are hard to reason about.

**Comment by swankjesse** [comment]
Thanks!
